examples/src/chains/map_reduce_lcel.ts

```diff
@@ -0,0 +1,163 @@
+  "Combine these summaries:\n\n{context}"
+);
+
+// Define a util function to allow function invocation with preset arguments
+function partial<F extends (...args: any[]) => any>(
+  fn: F,
+  ...partialArgs: any[]
+): (...args: any[]) => ReturnType<F> {
+  return (...args: any[]): ReturnType<F> => fn(...partialArgs, ...args);
+}
+
+// Wrap the `formatDocument` util so it can format a list of documents
+const formatDocs = async (documents: Document[]): Promise<string> => {
+  const formattedDocs = await Promise.all(
+    documents.map((doc) => formatDocument(doc, documentPrompt))
+  );
+  return formattedDocs.join("\n\n");
+};
+
+// Define a function to get the number of tokens in a list of documents
+const getNumTokens = async (documents: Document[]): Promise<number> =>
+  model.getNumTokens(await formatDocs(documents));
+
+// Initialize the output parser
+const outputParser = new StringOutputParser();
+
+// Define the map chain to format, summarize, and parse the document
+const mapChain = RunnableSequence.from([
+  { context: async (i: Document) => formatDocument(i, documentPrompt) },
+  summarizePrompt,
+  model,
+  outputParser,
+]);
+
+// Define a wrapper chain to keep the original metadata while summarizing the document
+const mapAsDocChain = RunnableSequence.from([
+  RunnableMap.from({ doc: new RunnablePassthrough(), content: mapChain }),
+  RunnableLambda.from(
+    (input) =>
+      new Document({ pageContent: input.content, metadata: input.doc.metadata })
+  ),
+]).withConfig({ runName: "Summarize (return doc)" });
+
+// Define the collapse chain to format, collapse, and parse a list of documents
+const collapseChain = RunnableSequence.from([
+  { context: async (i: { documents: Document[] }) => formatDocs(i.documents) },
+  collapsePrompt,
+  model,
+  outputParser,
+]);
+
+// Define a function to collapse a list of documents until the total number of tokens is within the limit
+const collapse = async (
+  documents: Document[],
+  config?: BaseCallbackConfig,
+  tokenMax = 4000
+) => {
+  const editableConfig = config;
+  let docs = documents;
+  let collapseCount = 1;
+  while ((await getNumTokens(docs)) > tokenMax) {
+    if (editableConfig) {
+      editableConfig.runName = `Collapse ${collapseCount}`;
+    }
+    const invoke = partial(collapseChain.invoke);
+    const splitDocs = _splitListOfDocs(docs, getNumTokens, tokenMax);
+    docs = await Promise.all(
+      splitDocs.map((doc) => _collapseDocs(doc, invoke))
+    );
+    collapseCount += 1;
+  }
+  return docs;
+};
+
+// Define the reduce chain to format, combine, and parse a list of documents
+const reduceChain = RunnableSequence.from([
+  { context: formatDocs },
+  combinePrompt,
+  model,
+  outputParser,
+]).withConfig({ runName: "Reduce" });
+
+// Define the final map-reduce chain
+const mapReduceChain = RunnableSequence.from([
+  mapAsDocChain.map(),
+  collapse,
+  reduceChain,
+]).withConfig({ runName: "Map reduce" });
```

Comment:
I would add something in the docs about the flow:

```
To show how this can be implemented in Expression Language, we break the map-reduce chain down into three steps:

1. Summarization of each document
2. Collapsing
3. Reducing
```

But also, don't we only summarize if the docs are smaller than the original context window?

Reply:
Updated mdx file with comments